pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,321 @@
+            densities.extend(new_densities)
+        rhs_int_gs.append(rhs_mat_int_gs[i][0].copy(
+            source_kernels=tuple(source_kernels), densities=tuple(densities)))
+
+    # Now that we have the IntG expressions depending on the source
+    # we now have to attach the target dependent derivatives.
+    res = [0]*lhs_mat.shape[0]
+    for i in range(lhs_mat.shape[0]):
+        for j in range(lhs_mat.shape[1]):
+            res[i] += _convert_target_poly_to_int_g(lhs_mat[i, j],
+                    int_gs[i], rhs_int_gs[j])
+
+    return res
+
+
+def _create_matrix(int_gs, source_dependent_variables, axis_vars):
+    """Create a matrix from a list of :class:`~pytential.symbolic.primitives.IntG`
+    objects and returns the matrix and the expressions corresponding to each column.
+    Each expression is one of ``source_dependent_variables`` or
+    equals to one. Each element in the matrix is a multi-variate polynomial
+    and the variables in the polynomial are from ``axis_vars`` input.
+    Each polynomial represents a derivative operator.
+    """
+    source_exprs = []
+    coefficient_collector = CoefficientCollector(source_dependent_variables)
+    to_sympy = PymbolicToSympyMapper()
+    matrix = []
+
+    for int_g in int_gs:
+        row = [0]*len(source_exprs)
+        for density, source_kernel in zip(int_g.densities, int_g.source_kernels):
+            d = coefficient_collector(density)
+            for source_expr, coeff in d.items():
+                if source_expr not in source_exprs:
+                    source_exprs.append(source_expr)
+                    row += [0]
+                poly = _convert_kernel_to_poly(source_kernel, axis_vars)
+                row[source_exprs.index(source_expr)] += poly * to_sympy(coeff)
+        matrix.append(row)
+
+    # At the beginning, we didn't know the number of columns of the matrix.
+    # Therefore we used a list for rows and they kept expanding.
+    # Here we are adding zero padding to make the result look like a matrix.
+    for row in matrix:
+        row += [0]*(len(source_exprs) - len(row))
+
+    return matrix, source_exprs
+
+
+def _check_int_gs_common(int_gs):
+    """Checks that the :class:`~pytential.symbolic.primtive.IntG` objects
+    have the same base kernel and other properties that would allow
+    merging them.
+    """
+    base_kernel = int_gs[0].source_kernels[0].get_base_kernel()
+    common_int_g = int_gs[0].copy(target_kernel=base_kernel,
+            source_kernels=(base_kernel,), densities=(1,))
+    for int_g in int_gs:
+        for source_kernel in int_g.source_kernels:
+            if source_kernel.get_base_kernel() != base_kernel:
+                return False
+        if common_int_g != int_g.copy(target_kernel=base_kernel,
+                source_kernels=(base_kernel,), densities=(1,)):
+            return False
+    return True
+
+
+def _convert_kernel_to_poly(kernel, axis_vars):
```

Comment:
```suggestion
def _kernel_target_derivs_as_poly(kernel, axis_vars):
```

Reply:
Fixed in 801e5fbdbd3e467d422fc9cd04907747bc074de6